snapcraft/plugins/kbuild.py

```diff
@@ -0,0 +1,181 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2016 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""The kbuild plugin is used for building kbuild based projects as snapcraft
+parts.
+
+The plugin applies your selected defconfig first by running
+
+    make defconfig
+
+and then uses the kconfigs flag to augment the resulting config by prepending
+the configured kconfigs values to the .config and running
+
+    "yes" "" | make oldconfig
+
+to create an updated .config file.
+
+If kconfigfile is provided this plugin will use the provided config file
+wholesale as the starting point instead of make $kdefconfig. In case user
+configures both a kdefconfig as well as kconfigfile, kconfigfile approach will
+be used.
+
+This plugin is based on the snapcraft.BasePlugin and supports the properties
+provided by that plus the following kbuild specific options with semantics as
+explained above:
+
+    - kdefconfig:
```

Comment:
I suggest reversing the order of listing the variables and the above description. Much of the above description mention the below variables by name but the reader has no context as to what those variables are until they get to the bottom.